### ISSUE TITLE
`geotargets` and cloud storage

### ISSUE THREAD

**Issue by Aariq** [njtierney/geotargets#112]
I *think* there may be a problem with using cloud storage.  When a SpatRaster is stored with `repository = "aws"`, for example, and a user tries to load it with `tar_read()` or `tar_load()`, I think what happens is the file is downloaded from AWS into `tempdir()/_targets/scratch`, then read in using the read function stored in `format`, then the file is deleted from the scratch dir once the target is loaded into memory.  This means the `SpatRaster` object makes it into memory, but the file it points to is gone. 

I've confirmed this behavior with a S3 bucket hosted on Jetstream2 and I can share (privately) the credentials for it if you'd like to test to confirm this.

I can't find an argument or option in `targets` that overrides this behavior, although I would have expected `memory = "persistent"` to maybe do something here.

```r
library(targets)
tar_dir({
  tar_script({
    library(targets)
    library(geotargets)
    tar_option_set(
      repository = "aws",
      resources = tar_resources(
        aws = tar_resources_aws(
          bucket = "test123456",
          prefix = "targets_test",
          endpoint = "https://js2.jetstream-cloud.org:8001"
        )
      )
    )
    list(
      tar_target(
        file,
        system.file("ex/elev.tif", package = "terra"),
        format = "file", 
        repository = "local"
      ),
      tar_terra_rast(
        rast_example,
        terra::rast(file)
      )
    )
  })
  tar_make()
  tar_load(rast_example)
  sources(rast_example)
  fs::file_exists(sources(rast_example))
})
#> ▶ dispatched target file
#> ● completed target file [0.003 seconds, 7.994 kilobytes]
#> ▶ dispatched target rast_example
#> ● completed target rast_example [0.008 seconds, 8.523 kilobytes]
#> ▶ ended pipeline [1.063 seconds]
#> /private/var/folders/wr/by_lst2d2fngf67mknmgf4340000gn/T/RtmpVmmt9h/_targets/scratch#> /rast_example1674d43dc896a 
#> FALSE
```

**Comment by Aariq** [comment]
This might not be something we can fix, but we should document it as a limitation and possibly open a discussion in the `targets` repo

**Comment by mdsumner** [comment]
Is the rast() call made with the prefix '/vsicurl' or '/vsis3/'? If it isn't GDAL will whole-download the file behind the scenes, which sounds like what you're seeing (I don't understand how to run your example above, it fails for me and I don't understand how to unpack it)

**Comment by Aariq** [comment]
Sorry, this was a bit of a "note-to-self" left on a Friday.  You'll need some credentials (which I could share privately) for that example to work.

The `rast()` call is not being made with /vsicurl/ or /vsis3/.  The way I *think* targets works with cloud storage is that when you run `tar_read()`/`tar_load()` or when a downstream target takes a cloud-stored target as a dependency, the target file is downloaded from the cloud into a scratch directory, read in using the read function stored in `tar_format()` (in this case, just `rast(path)`), and then apparently deleted from the scratch directory.

I don't know if it's possible to customize this behavior, but you're right that in this particular case it would make sense to be able to use `/vsis3/`.  I think there is likely something useful in this discussion about the new content addressable storage options in `targets`, in particular this comment (https://github.com/ropensci/targets/discussions/1232#discussioncomment-10277319) and `tar_repository_cas()` (https://github.com/ropensci/targets/blob/main/R/tar_repository_cas.R)

**Comment by Aariq** [comment]
This might just be a bug in `targets`—it seems reasonable for the `_scratch/` directory to stick around for at least the existence of the interactive R session.

**Comment by Aariq** [comment]
Opened discussion here: https://github.com/ropensci/targets/discussions/1367

**Comment by Aariq** [comment]
Changing the read function for `tar_terra_rast()` from  `function(path) terra::rast(path)` to `function(path) terra::rast(path) + 0` forces the `SpatRaster` source to be memory and then the targets work with cloud storage.  I'm not sure what the downsides of this are other than that in-memory `SpatRasters` appear to lose info about file blocksize (#99).

**Comment by Aariq** [comment]
Before adding some kind of "force_memory" option, the workaround I'd like to try and document as a vignette is to use the relatively new content addressable storage functionality in `targets` with a custom download function that doesn't actually download the raster, but returns a `/vsis3/bucket/key` style path that `terra::rast()` can read in.

**Comment by mdsumner** [comment]
maybe write to /vsimem and then ingest those bytes, that's what I've done with gdalraster (the VSIFile interface). I don't really see a problem with a tempfile though, but also can't terra write to /vsis3 ?  I'll try

**Comment by mdsumner** [comment]
the thing I'm looking at now (independent of terra) is wanting to track S3 objects that GDAL writes to "/vsis3" but targets then cannot understand as a file format, because (I think) targets didn't "put them there". 

so, while I had this equivalent to write to S3 (a todo of me to convert this to gdalraster, but terra could also do this)

```R
dsn <- "/vsicurl/https://projects.pawsey.org.au/idea-gebco-tif/GEBCO_2024.tif" 
ex <- c(-6007500, 6007500, -14422000, 14422000)
target_file <- "/vsis3/bucket/prefix/test.tif"
im <- vapour::gdal_raster_dsn(dsn,
                              target_res = 20000, target_crs = "+proj=tmerc +lon_0=115", target_ext = ex,
                              out_dsn = target_file)

```

we can convert that to the following

```R
temp_file <- "/vsimem/test.tif"
gdalraster::warp(dsn, temp_file, "+proj=tmerc +lon_0=115", cl_arg = c("-tr", 20000, 20000, "-te", ex[1], ex[3], ex[2], ex[4]))
vsids <- new(gdalraster::VSIFile, temp_file)
bytes <- vsids$ingest(-1)
vsids$close()
gdalraster::vsi_unlink(temp_file)
```

and, my questions then 

1.   how to have targets be responsible for those raw bytes at `target_file`?  I appreciate this is a targets question, not a geotargets one - but I think it's something we can use here (terra afaik doesn't expose VSIFile, but we could have it write to temp and ingest from that anyway (, which is what GDAL has to do for a GeoTIFF under the hood)). 

(This VSIFile approach is available in odc and uses rasterio for the underlying mech, but it's just a generic GDAL thing not rasterio)

2. Can we just use GDAL to write to the /vsi* location and then have targets take over?   (with 1. we have the problem of `bytes` being too big to slurp in one go, and GDAL already does multipart upload and that can be controlled so wbg to leverage that)

**Comment by brownag** [comment]
@mdsumner very thought provoking. I've been wanting to add some more geotargets tools for working with /vsi* paths and also gdalraster. Here are some initial ideas--there is a lot to unpack here. We might need to spin some of it off into other issues.

> the thing I'm looking at now (independent of terra) is wanting to track S3 objects that GDAL writes to "/vsis3" but targets then cannot understand as a file format, because (I think) targets didn't "put them there".

I think the main issue targets has with /vsi paths is no native way to parse the relevant URL or local file path from after the prefix. gdalraster provides a couple wrapper functions like `vsi_get_actual_url()` that could be used as part of a solution

> so, while I had this equivalent to write to S3 (a todo of me to convert this to gdalraster, but terra could also do this)
> 
> dsn <- "/vsicurl/https://projects.pawsey.org.au/idea-gebco-tif/GEBCO_2024.tif" 
> ex <- c(-6007500, 6007500, -14422000, 14422000)
> target_file <- "/vsis3/bucket/prefix/test.tif"
> im <- vapour::gdal_raster_dsn(dsn,
>                               target_res = 20000, target_crs = "+proj=tmerc +lon_0=115", target_ext = ex,
>                               out_dsn = target_file)
> we can convert that to the following
> 
> temp_file <- "/vsimem/test.tif"
> gdalraster::warp(dsn, temp_file, "+proj=tmerc +lon_0=115", cl_arg = c("-tr", 20000, 20000, "-te", ex[1], ex[3], ex[2], ex[4]))
> vsids <- new(gdalraster::VSIFile, temp_file)
> bytes <- vsids$ingest(-1)
> vsids$close()
> gdalraster::vsi_unlink(temp_file)
> and, my questions then


> 1. how to have targets be responsible for those raw bytes at `target_file`?  I appreciate this is a targets question, not a geotargets one - but I think it's something we can use here (terra afaik doesn't expose VSIFile, but we could have it write to temp and ingest from that anyway (, which is what GDAL has to do for a GeoTIFF under the hood)).

tarchetypes has `tar_files()` for local and remote files outside the target store. It has a few formats, corresponding to different ways of evaluating outdated-ness of the file. There is a concept for "aws_file" format, but it has no understanding of GDAL  /vsi* paths, so does not work the way we want

I imagine geotargets could provide the GDAL analogue: something like `tar_vsi_files()` and a "vsi_file" format. 

This function would operate similarly to `tar_files()` as a 2-target factory with an "upstream" target corresponding to the input (VSI) path, and a downstream target corresponding to a hash or other information providing info on version of that remote source. Depending on the source and protocol being used, there would probably need to be a few options (as with `tar_files()`). The vsi_file format would recognize that using GDAL we can create a spatraster object (or any other GDAL-based command) that can internally contain that /vsi path as a lightweight reference to the data stored remotely. Provided we have a target to check that the remote data are up to date at that path, the local target store would need only contain that reference URL, from which we can reconstruct a SpatRaster, or a VSIFile, or just a raw "gdal" command on the fly in the format's custom "read" method. 

We could provide a helper method `extract_vsi_path()` or similar to convert vsi_files object to character path that can be passed to GDAL. Perhaps we could also have other options for parsing the path to extract information about user-supplied VSI paths, as we would need to keep track of the VSI prefixes needed on a per-target basis 

> 2. Can we just use GDAL to write to the /vsi* location and then have targets take over?   (with 1. we have the problem of `bytes` being too big to slurp in one go, and GDAL already does multipart upload and that can be controlled so wbg to leverage that)

Here is an example that can be done with current geotargets... but it is not very "pretty"... some convenience methods would help. This uses terra for warping for sake of an example and functions we already have target factories for. Using metadata obtained via gdalraster to check whether it has changed and whether the `tar_terra_rast()` (or some other raster-generating target) needs to run.

``` r
library(targets)

tar_script({
    library(geotargets)
    library(terra)
    list(
        # first two targets: candidate for `tar_vsi_files()` helper
        tar_target(
            x, "/vsicurl/https://projects.pawsey.org.au/idea-gebco-tif/GEBCO_2024.tif"
        ),
        tar_target(x_hash, {
            gdalraster::vsi_get_file_metadata(x, "headers")[
                c("etag", "content-length", "last-modified")
            ]
        }, cue = tar_cue(mode = "always")),
        tar_target(extent, c(-6007500, 6007500, -14422000, 14422000)),
        tar_target(resolution, 20000),
        tar_target(crs, "+proj=tmerc +lon_0=115"),
        tar_terra_rast(result, {
            x_hash # explicit dependency on the "hash"
            rt <- rast(extent = extent, res = resolution, crs = crs)
            project(rast(x), rt, align_only = TRUE, use_gdal = TRUE, by_util = TRUE)
        })
    )
})

# tar_destroy()
tar_make()
#> terra 1.8.80
#> + x dispatched
#> ✔ x completed [0ms, 121 B]
#> + resolution dispatched
#> ✔ resolution completed [0ms, 53 B]
#> + crs dispatched
#> ✔ crs completed [0ms, 75 B]
#> + extent dispatched
#> ✔ extent completed [0ms, 63 B]
#> + x_hash dispatched
#> ✔ x_hash completed [1.8s, 193 B]
#> + result dispatched
#> GDAL WARNING 6: warp options does not support option -r
#> ✔ result completed [21.5s, 8.47 MB]
#> ✔ ended pipeline [23.8s, 6 completed, 0 skipped]

tar_load(x_hash)
x_hash
#> $etag
#> [1] "\"4893b12085a7a7b8b35aa0b2a062853e-549\""
#> 
#> $`content-length`
#> [1] "4598597308"
#> 
#> $`last-modified`
#> [1] "Tue, 18 Feb 2025 01:30:44 GMT"

tar_load(result)
result
#> class       : SpatRaster 
#> size        : 1992, 1671, 1  (nrow, ncol, nlyr)
#> resolution  : 20000, 20000  (x, y)
#> extent      : -16707500, 16712500, -19822000, 20018000  (xmin, xmax, ymin, ymax)
#> coord. ref. : +proj=tmerc +lat_0=0 +lon_0=115 +k=1 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs 
#> source      : result 
#> name        :  lyr.1 
#> min value   : -10456 
#> max value   :   6400

terra::plot(result)
```

![](https://i.imgur.com/0m7xHv8.png)

This method does not particularly care _how_ the remote file got there, just what its metadata (etag, last modified, whathaveyou) are to know whether downstream targets are up to date. 

Ideally, getting back to @Aariq's OP, we could make it so that we could write targets to e.g. S3, Azure, Google Cloud files by setting the appropriate `tar_resources()`... and then access those assets automagically via geotargets methods constructing a "read" with with corresponding /vsi* prefix. Essentially, a way of bypassing the issues with the scratch files (downloaded locally only to load in memory and delete) by just returning a user level object or command with the remote path

**Comment by mdsumner** [comment]
Nice!  I have to read this in detail, but glad to have this response. `tar_files()` might do what I want, and I didn't know about `vsi_get_actual_url()`. 

To update what I was looking for in previous post: 

TLDR: I have a lot to explore from this response and I'm still in the middle of stuff I've learnt since. 

I realized with feedback (https://github.com/ropensci/targets/discussions/852#discussioncomment-14933354) that what I expected from `target = "file"` with an aws store is not possible. It seems like it might be with CAS, but for now I have stopped trying to track the s3:// addresses and I force it to run on every input, but shortcut if the s3:// address already exists (I'm still mixing s3:// and /vsis3 with gsub in here because I'm not sure what I want yet). This gives the peformance that I want, but takes 2 minutes or so while every s3:// address is 'stat' checked. 

https://github.com/mdsumner/seaice.map/blob/74f04448b64d3441c84ff6f73085d0c387d1b47c/_targets.R#L47-L50

It means that no source will invalidate if the source URL changes, though.  FWIW, I don't really like working with s3:// or /vsis3 because, apart from Amazon-native stores, you always need the user to set some config so you don't have that light flexiblity anymore.  

(I have a mix of bowerbird and targets in that project, and we're discussing how bowerbird might align with targets in the longer term since they overlap quite a bit.  We've added a lot better capability for bowerbird to just get a list of urls, without actually copying them locally or to another store).
